python/lsst/dm/OCPS/ocps_csc.py

```diff
@@ -110,38 +144,111 @@ def __init__(
+                    include=[trigger.event],
+                )
+                event = remote.getattr(trigger.event)
+                event.callback = self.gen_event_callback(trigger)
+                self.trigger_remotes.append(remote)
+
+    def gen_event_callback(self, trigger):
+        """Return a callback that triggers a pipeline on an event.
+
+        Parameters
+        ----------
+        trigger: types.SimpleNamespace
+            Must contain version, pipeline, data_query_expr attributes
+
+        Notes
+        -----
+        version: str
+            Science Pipelines version as an EUPS tag
+        pipeline: str
+            URL of pipeline YAML
+        data_query_expr: str
+            String to generate a data query expression for "pipetask run",
+            formatted with ``.format(data=data)`` so ``{data.attr}``
+            substitutions can be used
+        """
+        async def event_callback(self, data):
+            self.log.info(f"Event {trigger.event} occurred: {data}")
+            data.version = trigger.version
+            data.pipeline = trigger.pipeline
+            data.config = ""
+            data.data_query = trigger.data_query_expr.format(event=data)
+            self.log.info(f"Calling _execute with {data}")
+            # TODO DM-30032: complete event-triggered execution
+            # self._execute(data)
+
+        return event_callback
 
     async def do_execute(self, data):
         """Implement the ``execute`` command.
 
-        Submit a request for execution to the back-end REST API.
-
         Parameters
         ----------
         data: types.SimpleNamespace
-            Must contain pipeline, version, and config attributes
+            Must contain version, pipeline, config, and data_query attributes
         """
         self.assert_enabled("execute")
         self.log.info(f"execute command with {data}")
+        await self._execute(data)
+
+    async def _execute(self, data):
+        """Submit a request for execution to the back-end REST API.
+
+        Parameters
+        ----------
+        data: types.SimpleNamespace
+            Must contain version, pipeline, config, data_query attributes
+
+        Notes
+        -----
+        version: str
+            Science Pipelines version as an EUPS tag
+        pipeline: str
+            URL of pipeline YAML
+        config: str
+            Command line options for "pipetask run"
+        data_query: str
+            Data query expression for "pipetask run"
 
+        The ``data.config`` attribute is expected to contain ``-i`` options
+        beyond the default OODS input collection and ``-c`` options to
+        override configuration values.
+        """
         if self.simulation_mode == 0:
             # Real command.
+            run_options = ""
+            if hasattr(self.config, "input_collection"):
+                run_options = f"-i {self.config.input_collection}"
+            payload_env = dict(
```

Comment:
I'm wondering about using pydantic at this point.

Reply:
Russell already has a whole JSON Schema thing going.